### PR TITLE
fix: Exchange rate precompile return error when called with value #13052

### PR DESCRIPTION
**Pull request by stefan-stefanooov**
**Description**:
Executing an internal call with value against PRNG/Exchange rate precompile addresses returns success. Only the contract that acts as middleman is credited with hbars.

**Related issue(s)**:
https://github.com/hashgraph/hedera-services/pull/12793

Fixes #https://github.com/hashgraph/hedera-services/issues/11218

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java [hunk 1]
```diff
@@ -100,6 +103,7 @@ void convertsPositiveNumberToTinycentsAsExpected() {
 
     @Test
     void convertsZeroToTinybarsAsExpected() {
+        given(frame.getValue()).willReturn(Wei.ZERO);
         givenRate(someRate);
 
         final var zeroInput = tinycentsInput(0);
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java:109]: Should there be a unit test that tests the new exception case?

**Review by lukelee-sl — COMMENTED** [review]
Seems reasonable but should there not be a HAPI test to show that the transaction reverts as outlined in the original issue?

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java (hunk 1 above)
Comment (stefan-stefanooov) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java:109]: Added.

**Review by lukelee-sl — DISMISSED** [review]
LGTM

**Comment by stefan-stefanooov** [comment]
> Seems reasonable but should there not be a HAPI test to show that the transaction reverts as outlined in the original issue?

Hey @lukelee-sl 
Thank you for the review. 
I have added integration test as well.
Please have another look.

File: hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java [hunk 2]
```diff
@@ -118,6 +122,17 @@ void inputCannotUnderflow() {
...
 
+    @Test
+    void valueShouldNotBeSentToThePrecompile() {
+        given(frame.getValue()).willReturn(Wei.MAX_WEI);
+
+        final var zeroInput = tinycentsInput(0);
+        final var result = subject.computeFully(zeroInput, frame);
+
+        assertThat(result.output()).isEqualTo(Bytes.EMPTY);
+        assertThat(result.result().getHaltReason().get()).isEqualTo(ExceptionalHaltReason.INVALID_OPERATION);
```
Comment (petreze) [inline hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java:133]: Shouldn't we test for the new `INVALID_FEE_SUBMITTED` case?

Comment (stefan-stefanooov) [reply hedera-node/hedera-smart-contract-service-impl/src/test/java/com/hedera/node/app/service/contract/impl/test/exec/systemcontracts/ExchangeRateSystemContractTest.java]: A new version available. Thank you.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/PrngSystemContract.java [hunk 3]
```diff
@@ -102,7 +102,7 @@ public FullResult computeFully(@NonNull final Bytes input, @NonNull final Messag
                     PrecompiledContract.PrecompileContractResult.halt(Bytes.EMPTY, Optional.of(INVALID_OPERATION)),
                     gasRequirement,
                     null);
-        } catch (NullPointerException e) {
+        } catch (Exception e) {
```
Comment (stefan-stefanooov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/PrngSystemContract.java:105]: Not in the scope, but in the current flow we handle only two types of exceptions. If we have something throw another  exception type, we are going to leak it out. 
Putting the most common type at the end of the `catch` chain, we ensure that all exceptions are handled.

Discussed with the team. 
Any comments?

**Comment by lukelee-sl** [comment]
> > Seems reasonable but should there not be a HAPI test to show that the transaction reverts as outlined in the original issue?
> 
> Hey @lukelee-sl Thank you for the review. I have added integration test as well. Please have another look.

Thanks for adding the test. Looks good.

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Review by petreze — APPROVED** [review]

**Pull request merged**